fastar/update_rules.py

```diff
@@ -21,66 +22,147 @@ def take(arr, idxs):
+                                        zip(output_slice[-np.ndim(b):], np.shape(b)))))
+
 
 def dot_update(old_out, a, b):
     a, a_mask = a
     b, b_mask = b
-    outval, old_outmask = old_out[0] if old_out else _init_out(lax.dot_p, a, b)
-    outmask = onp.equal(onp.dot(a_mask.astype(int), b_mask.astype(int)), onp.shape(b_mask)[0])
-    new_mask = outmask & ~old_outmask
-    out_slices = util.mask_to_slices(new_mask)
-    for slc in out_slices:
-        a_slice = (slc[0], slice(None, None)) if len(slc) > 0 else (slice(None, None),)
-        b_slice = (slice(None, None), slc[1]) if len(slc) > 1 else (slice(None, None),)
-        assert np.all(outval[slc] == 0)
-        outval = _add_at(outval, slc, lax.dot_p.bind(a[a_slice], b[b_slice]))
-    return fa.Parray((outval, outmask))
-
-unary_functions = [lax.sin_p]
-for f in unary_functions:
-    fa.update_rules[f] = unop_update(f)
-
-binary_functions = [lax.add_p, lax.sub_p, lax.mul_p]
-for f in binary_functions:
-    fa.update_rules[f] = binop_update(f)
+    return sliceableop_update(
+        lax.dot_p, old_out, a, b, output_mask=onp.equal(
+            onp.dot(a_mask.astype(int), b_mask.astype(int)), onp.shape(b_mask)[0]),
+        input_slices_from_output_slice=lambda output_slice: (
+            (output_slice[0], all) if len(output_slice) > 0 else (all,),
+            (all, output_slice[1]) if len(output_slice) > 1 else (all,)))
+
+
+unops = [lax.sin_p]
+for op in unops:
+    fa.update_rules[op] = unop_update(op)
+
+reduce_ops = [lax.reduce_sum_p, lax.reduce_min_p, lax.reduce_max_p]
+for op in reduce_ops:
+    fa.update_rules[op] = reduce_update(op)
+
+binops = [lax.add_p, lax.sub_p, lax.mul_p]
+for op in binops:
+    fa.update_rules[op] = binop_update(op)
 
 fa.update_rules[lax.dot_p] = dot_update
 
+
+def pad_update(old_out, a, padding_value, padding_config):
+    a, a_mask = a
+    # TODO Support unitialized padding_value.
+    # Is currently assumed to be initialized from the start, mask is ignored:
+    padding_value, _ = padding_value
+
+    def init_output(func, *args, **params):
+        outval, outmask = _init_output(func, *args, **params)
+
+        for index, (lo, hi, interior) in enumerate(padding_config):
+            # TODO:
+            assert interior == 0, "Interior padding not yet supported, expected 0."
+
+            lower_slice = [slice(None, lo) if index == i else all
+                           for i, s in enumerate(outmask.shape)]
+            higher_slice = [slice(-hi, None) if index == i else all
+                            for i, s in enumerate(outmask.shape)]
+
+            outmask[lower_slice] = True
+            outmask[higher_slice] = True
+
+        return outval, outmask
+
+    output_mask = onp.ones(tuple(s + lo + hi for s, (lo, hi, _) in zip(a.shape, padding_config)),
+                           dtype=bool)
+    output_mask[tuple([slice(lo, -hi, None) for lo, hi, _ in padding_config])] = a_mask
```

Comment:
Can’t the output mask be computed easily by calling pad on the input mask, with the pad value set to True or False depending on the input pad value’s mask?

Reply:
Good point, absolutely